### PR TITLE
[fix](hudi) move wrong members in HMSExternalTable

### PR DESCRIPTION
**Pull request by morningman**
Previously, there are 2 members: TableScanParams and IncrementalRelation in HMSExternalTable.
These 2 members are for Hudi's incremental query, so their lifecycle should be with query task,
should not be saved in HMSExternalTable.

This PR mainly changes:

- Add LogicalHudiScan and PhysicalHudiScan, extends from LogicalFileScan and PhysicalFileScan.
- Move TableScanParams and IncrementalRelation from HMSExternalTable to XXXHudiScan.
- Add or modify related Nereids rules

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/RelationVisitor.java [hunk 1]
```diff
@@ -154,6 +160,10 @@ default R visitPhysicalFileScan(PhysicalFileScan fileScan, C context) {
         return visitPhysicalCatalogRelation(fileScan, context);
     }
 
+    default R visitPhysicalHudiScan(PhysicalHudiScan fileScan, C context) {
+        return visitPhysicalCatalogRelation(fileScan, context);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/RelationVisitor.java:164]: if it extends LogicalFileScan, it should
`return visitPhysicalFileScan(fileScan, context);`

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/RelationVisitor.java [hunk 2]
```diff
@@ -154,6 +160,10 @@ default R visitPhysicalFileScan(PhysicalFileScan fileScan, C context) {
         return visitPhysicalCatalogRelation(fileScan, context);
     }
 
+    default R visitPhysicalHudiScan(PhysicalHudiScan fileScan, C context) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/RelationVisitor.java:163]:
```suggestion
    default R visitPhysicalHudiScan(PhysicalHudiScan hudiScan, C context) {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/RelationVisitor.java [hunk 3]
```diff
@@ -105,6 +107,10 @@ default R visitLogicalFileScan(LogicalFileScan fileScan, C context) {
         return visitLogicalExternalRelation(fileScan, context);
     }
 
+    default R visitLogicalHudiScan(LogicalHudiScan fileScan, C context) {
+        return visitLogicalExternalRelation(fileScan, context);
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/RelationVisitor.java:111]: same as physcial visitor

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalHudiScan.java [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Set;
+
+/**
+ * Physical Hudi scan for Hudi table.
+ */
+public class PhysicalHudiScan extends PhysicalFileScan {
+
+    // for hudi incremental read
+    private Optional<TableScanParams> scanParams;
+    private Optional<IncrementalRelation> incrementalRelation;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalHudiScan.java:48]: final

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalHudiScan.java [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    public PhysicalHudiScan(RelationId id, ExternalTable table, List<String> qualifier,
+            DistributionSpec distributionSpec, Optional<GroupExpression> groupExpression,
+            LogicalProperties logicalProperties, Set<Expression> conjuncts,
+            SelectedPartitions selectedPartitions, Optional<TableSample> tableSample,
+            Optional<TableSnapshot> tableSnapshot,
+            Optional<TableScanParams> scanParams, Optional<IncrementalRelation> incrementalRelation) {
+        super(id, PlanType.PHYSICAL_HUDI_SCAN, table, qualifier, distributionSpec, groupExpression, logicalProperties,
+                conjuncts, selectedPartitions, tableSample, tableSnapshot);
+        this.scanParams = scanParams;
+        this.incrementalRelation = incrementalRelation;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalHudiScan.java:62]: check not null

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHudiScan.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+/**
+ * Logical Hudi scan for Hudi table
+ */
+public class LogicalHudiScan extends LogicalFileScan {
+    private static final Logger LOG = LogManager.getLogger(LogicalHudiScan.class);
+
+    // for hudi incremental read
+    private Optional<TableScanParams> scanParams;
+    private Optional<IncrementalRelation> incrementalRelation;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHudiScan.java:65]: final

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHudiScan.java [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                Optional.empty(), Optional.empty());
+    }
+
+    /**
+     * Set scan params for incremental read
+     *
+     * @param table
+     * @param scanParams
+     */
+    public void setScanParams(HMSExternalTable table, TableScanParams scanParams) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHudiScan.java:94]: should not do set in Nereids, if u want to update some attr, u should impl withXXX and return a new object

File: fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java [hunk 8]
```diff
@@ -465,6 +472,11 @@ public Statistics visitPhysicalFileScan(PhysicalFileScan fileScan, Void context)
         return computeCatalogRelation(fileScan);
     }
 
+    @Override
+    public Statistics visitPhysicalHudiScan(PhysicalHudiScan fileScan, Void context) {
+        return computeCatalogRelation(fileScan);
+    }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/stats/StatsCalculator.java:478]: remove this code if it same with PhysicalFileScan

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHudiScan.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            Optional<GroupExpression> groupExpression, Optional<LogicalProperties> logicalProperties,
+            Set<Expression> conjuncts, SelectedPartitions selectedPartitions, Optional<TableSample> tableSample,
+            Optional<TableSnapshot> tableSnapshot,
+            Optional<TableScanParams> scanParams, Optional<IncrementalRelation> incrementalRelation) {
+        super(id, table, qualifier, groupExpression, logicalProperties, conjuncts,
+                selectedPartitions, tableSample, tableSnapshot);
+        Objects.requireNonNull(scanParams);
+        Objects.requireNonNull(incrementalRelation);
+        this.scanParams = scanParams;
+        this.incrementalRelation = incrementalRelation;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHudiScan.java:81]:
```suggestion
        ;
        ;
        this.scanParams = Objects.requireNonNull(scanParams, "scanParams should not null");
        this.incrementalRelation = Objects.requireNonNull(incrementalRelation, "incrementalRelation should not null");
```

**Comment by morningman** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalHudiScan.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+            LogicalProperties logicalProperties, PhysicalProperties physicalProperties,
+            Statistics statistics, Set<Expression> conjuncts, SelectedPartitions selectedPartitions,
+            Optional<TableSample> tableSample, Optional<TableSnapshot> tableSnapshot,
+            Optional<TableScanParams> scanParams, Optional<IncrementalRelation> incrementalRelation) {
+        super(id, PlanType.PHYSICAL_HUDI_SCAN, table, qualifier, distributionSpec, groupExpression, logicalProperties,
+                physicalProperties, statistics, conjuncts, selectedPartitions, tableSample, tableSnapshot);
+        this.scanParams = scanParams;
+        this.incrementalRelation = incrementalRelation;
+    }
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalHudiScan.java:83]: equals and hashCode?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHudiScan.java [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+                Optional.empty(), conjuncts, selectedPartitions, tableSample, tableSnapshot,
+                scanParams, incrementalRelation);
+    }
+
+    @Override
+    public <R, C> R accept(PlanVisitor<R, C> visitor, C context) {
+        return visitor.visitLogicalHudiScan(this, context);
+    }
+
+    /**
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalHudiScan.java:175]: equals and hashCode?

**Comment by morningman** [comment]
run buildall

**Comment by morningman** [comment]
run buildall

**Comment by AshinGau** [comment]
LGTM

**Review by AshinGau — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**